pkg/cmd/cli/config/store.go

```diff
@@ -0,0 +1,165 @@
+	Path           string
+	providerEngine string
+}
+
+type ConfigStoreNotFoundError struct {
+	msg string
+}
+
+func (e *ConfigStoreNotFoundError) Error() string {
+	return e.msg
+}
+
+func (c *ConfigStore) FromOpenShift() bool {
+	return c.providerEngine == fromOpenShift
+}
+
+func (c *ConfigStore) FromKube() bool {
+	return c.providerEngine == fromKube
+}
+
+// Load a ConfigStore from the explicit path to a config file provided as argument
+// Error if not found.
+func LoadFrom(path string) (*ConfigStore, error) {
+	config, err := tryToLoad(path, fromOpenShift)
+	if err == nil {
+		return config, nil
+	}
+
+	return nil, &ConfigStoreNotFoundError{fmt.Sprintf("Unable to load config file from '%v': %v", path, err.Error())}
+}
+
+// Load a ConfigStore using the priority conventions declared by the ClientConfigLoadingRules.
+// Error if none can be found.
+func LoadWithLoadingRules() (store *ConfigStore, err error) {
+	for _, path := range OpenShiftClientConfigFilePriority() {
+		store, err = tryToLoad(path, fromOpenShift)
+		if err == nil {
+			return store, nil
+		}
+	}
+
+	for _, path := range KubeClientConfigFilePriority() {
+		store, err = tryToLoad(path, fromKube)
+		if err == nil {
+			return store, nil
+		}
+	}
+
+	return nil, &ConfigStoreNotFoundError{fmt.Sprintf("Unable to load config file from any of the expected locations: %v", err.Error())}
+}
+
+// Create a new file to store configs and returns the ConfigStore that represents it.
+func CreateEmpty() (*ConfigStore, error) {
+	configPathToCreateIfNotFound := fmt.Sprintf("%v/%v", os.Getenv("HOME"), OpenShiftConfigHomeDirFileName)
+
+	glog.V(3).Infof("A new config will be created at: %v ", configPathToCreateIfNotFound)
+
+	newConfig := clientcmdapi.NewConfig()
+
+	if err := os.MkdirAll(fmt.Sprintf("%v/%v", os.Getenv("HOME"), OpenShiftConfigHomeDir), 0755); err != nil {
+		return nil, fmt.Errorf("Tried to create a new config file but failed while creating directory %v: %v", OpenShiftConfigHomeDirFileName, err)
+	}
+	glog.V(5).Infof("Created directory %v", "~/"+OpenShiftConfigHomeDir)
+
+	if err := clientcmd.WriteToFile(*newConfig, configPathToCreateIfNotFound); err != nil {
+		return nil, fmt.Errorf("Tried to create a new config file but failed with: %v", err)
+	}
+	glog.V(5).Infof("Created file %v", configPathToCreateIfNotFound)
+
+	config, err := tryToLoad(configPathToCreateIfNotFound, fromOpenShift)
+	if err != nil {
+		return nil, err
+	}
+
+	return config, nil
+}
+
+// Save the provided config attributes to this ConfigStore.
+func (c *ConfigStore) SaveToFile(username, token string, namespace string, clientCfg *client.Config, rawCfg clientcmdapi.Config) error {
+	glog.V(4).Infof("Trying to merge and update %v config to '%v'...", c.providerEngine, c.Path)
+
+	config := clientcmdapi.NewConfig()
+
+	credentialsName := username
+	if len(credentialsName) == 0 {
+		credentialsName = "osc-login"
+	}
+	credentials := clientcmdapi.NewAuthInfo()
+	credentials.Token = token
```

Comment:
The server might be using client certs to secure the connection (clients without the right cert can't even establish a TLS connection), and the token for identity (GCE does this).